draft-ietf-quic-http.md

```diff
@@ -430,29 +469,79 @@ The flags defined are:
     1 and 256.
 
-A PRIORITY frame MAY identify a dependent stream with a stream ID of 0; as in
+A PRIORITY frame identifies a request to priotize, and a request upon which that
+request is dependent.  A request is identified by the stream ID of a request
+when the corresponding PUSH_PRIORITIZED or PUSH_DEPENDENT flag is not set.
+Setting the PUSH or PUSH_DEPENDENT flag causes the frame to identify a
+PUSH_PROMISE using a Push ID (see {{frame-push-promise}} for details).
+
+A PRIORITY frame MAY identify a Stream Dependency with a stream ID of 0; as in
 {{!RFC7540}}, this makes the request dependent on the root of the dependency
-tree.  Stream ID 0 and stream ID 1 cannot be reprioritized; an attempt to
-reprioritize these stream MUST be treated as a connection error of type
+tree.
+
+Stream ID 0 and stream ID 1 cannot be reprioritized. A Prioritized Request that
+identifies Stream 0 or 1 MUST be treated as a connection error of type
 HTTP_MALFORMED_PRIORITY.
 
+A PRIORITY frame that does not reference a request MUST be treated as a
+HTTP_MALFORMED_PRIORITY error, unless it references stream ID 0.  A PRIORITY
+that sets a PUSH_PRIORITIZED or PUSH_DEPENDENT flag, but then references a
+non-existent Push ID MUST be treated as a HTTP_MALFORMED_PRIORITY error.
+
 The length of a PRIORITY frame is 9 octets.  A PRIORITY frame with any other
 length MUST be treated as a connection error of type HTTP_MALFORMED_PRIORITY.
 
 
+### CANCEL_PUSH {#frame-cancel-push}
+
+The CANCEL_PUSH frame (type=0x3) is used to request cancellation of server push
+prior to the push stream being created.  The CANCEL_PUSH frame identifies a
+server push request by Push ID (see {{frame-push-promise}}).
+
+When a server receives this frame, it aborts sending the response for the
+identified server push.  If the server has not yet started to send the server
+push, it can use the receipt of a CANCEL_PUSH frame to avoid opening a
+stream.  If the push stream has been opened by the server, the server SHOULD
+sent a QUIC RST_STREAM frame on those streams and cease transmission of the
+response.
+
+A server can send this frame to indicate that it won't be sending a response
+prior to creation of a push stream.  Once the push stream has been created,
+sending CANCEL_PUSH has no effect on the state of the push stream.  A QUIC
+RST_STREAM frame SHOULD be used instead to cancel transmission of the server
+push response.
+
+A CANCEL_PUSH frame is sent on the control stream.  Sending a CANCEL_PUSH frame
+on a stream other than the control stream MUST be treated as a stream error of
+type HTTP_WRONG_STREAM.
+
+The CANCEL_PUSH frame has no defined flags.
+
+The CANCEL_PUSH frame carries a 32-bit Push ID that identifies the server push
+that is being cancelled (see {{frame-push-promise}}).
+
+A server MUST treat a CANCEL_PUSH frame payload that is other than 4 octets in
+length, or a CANCEL_PUSH frame that identifies an unknown Push ID as a
+connection error of type HTTP_MALFORMED_CANCEL_PUSH.
```

Comment:
A CANCEL_PUSH from a server could be reordered with a PUSH_PROMISE... I'm not sure what to do about this but I wanted to raise it.